piptools/utils.py

```diff
@@ -115,7 +116,16 @@ def format_requirement(
     if ireq.editable:
         line = f"-e {ireq.link.url}"
     elif is_url_requirement(ireq):
-        line = ireq.link.url
+        # if the requirement has no name then it's just a URL
+        if not ireq.name:
+            line = ireq.link.url
+        # otherwise parse the URL
+        else:
+            parsed_url = urlparse(ireq.link.url)
+            if "egg=" in parsed_url.fragment:
```

Comment:
```suggestion
            if ireq.link.egg_fragment:
```

`egg_fragment` can be used

Reply:
Nice catch!